src/main/go/RelpConnection.go

```diff
@@ -0,0 +1,305 @@
+
+		sendErr := relpConn.SendRelpRequest(relpRequest)
+		if sendErr != nil {
+			log.Printf("Error sending relp request: '%v'\n", err.Error())
+		}
+
+		ackErr := relpConn.ReadAcks(batch)
+		if ackErr != nil {
+			// ACK timeout or other failure
+			return ackErr
+		}
+	}
+
+	return nil
+}
+
+// ReadAcks reads the ACKs from the given batch.
+func (relpConn *RelpConnection) ReadAcks(batch *RelpBatch) error {
+	log.Printf("ReadAcks.Entry> Reading ACKs for batchID: %v\n", batch.requestId)
+	var parser *RelpParser = nil
+	notComplete := relpConn.window.Size() > 0
+	var cn = *relpConn.connection
+
+	for notComplete { // until window is empty
+		readBytes := 0
+		for { // until parse complete
+			if parser == nil {
+				parser = &RelpParser{}
+			}
+
+			// set ACK timeout duration, default 30 sec
+			errDl := cn.SetReadDeadline(time.Now().Add(relpConn.ackTimeoutDuration))
+			if errDl != nil {
+				return errors.New("error setting connection timeout")
+			}
+
+			n, err := cn.Read(relpConn.preAllocRxBuffer)
+			if err != nil {
+				if errors.Is(err, os.ErrDeadlineExceeded) {
+					// reading timed out
+					return &AckReadingError{reason: "timeout"}
+				} else if err == io.EOF {
+					return &AckReadingError{reason: "eof"}
+				} else {
+					// other error
+					return &AckReadingError{reason: "unexpected error: " + err.Error()}
+				}
+			} else {
+				readBytes += n
+			}
+
+			// parse all bytes in buffer
+			for i := 0; i < n; i++ {
+				parseErr := parser.Parse(relpConn.preAllocRxBuffer[i])
+				if parseErr != nil {
+					panic("parsing error: " + parseErr.Error())
+				}
+			}
+
+			if parser.isComplete {
+				log.Printf("ReadAcks> Parsing complete, with %v byte(s) read\n", readBytes)
+				// resp read successfully
+				txnId := parser.frameTxnId
+				if relpConn.window.IsPending(txnId) {
+					reqId, err := relpConn.window.GetPending(txnId)
+					if err != nil {
+						panic("Could not find given pending txnId from RelpWindow!")
+					}
+					response := RelpFrameRX{
+						RelpFrame{
+							transactionId: parser.frameTxnId,
+							cmd:           parser.frameCmdString,
+							dataLength:    parser.frameLen,
+							data:          parser.frameData.Bytes(),
+						},
+					}
+					batch.PutResponse(reqId, &response)
+					relpConn.window.RemovePending(txnId)
+				}
+
+				parser = nil
+				if relpConn.window.Size() == 0 {
+					// window empty, can exit readAcks method
+					notComplete = false
+				}
+				break
+			}
+		}
+
+	}
+	log.Println("ReadAcks.Done> Return with no errors")
+	return nil
+}
+
+// SendRelpRequest sends the RELP frame to the connected RELP server
+func (relpConn *RelpConnection) SendRelpRequest(tx *RelpFrameTX) error {
+	txN, err := tx.Write(relpConn.preAllocTxBuffer)
+
+	var cn = *relpConn.connection
+	dlErr := cn.SetWriteDeadline(time.Now().Add(relpConn.writeTimeoutDuration))
+	if dlErr != nil {
+		return dlErr
+	}
+
+	n, err := cn.Write(relpConn.preAllocTxBuffer.Bytes())
```

Comment:
how do this and L287 work together, especially if tx is many times larger than the preAllocTxBuffer?
Java version checks if the buffer is sufficent in size and allocates new one if it exceeds the existing buffer.

In C it is possible to do partial writes to a buffer in a **loop** so that it's written completely, eventually, by using a single buffer.

Reply:
the bytes.Buffer's internal byte array grows in size if the bytes written exceed the initial buffer size.
The size given in L45 is only the initial size.

Reply:
> how do this and L287 work together, especially if tx is many times larger than the preAllocTxBuffer? Java version checks if the buffer is sufficent in size and allocates new one if it exceeds the existing buffer.
> 
> In C it is possible to do partial writes to a buffer in a **loop** so that it's written completely, eventually, by using a single buffer.

the bytes.Buffer's internal byte array grows in size if the bytes written exceed the initial buffer size.
The size given in L45 is only the initial size.